### ISSUE TITLE
Thoughts around adding features to NER model

### ISSUE THREAD

**Issue by rbhambriiit** [explosion/spaCy#2328]
## Feature description
NER model to use parser, tagger features.

Background:
I have watched Spacy's video to get an idea on how NER model is trained based on Embed, Encode, Attend, Predict.

I have trained custom blank NER models on my dataset.

Idea:
The NER model should rely on other pipelines such as parser, tagger.
Features for pos, dependency label, previous-word-pos etc. should be considered at the embed stage.

I would like to hear more from @honnibal if this makes sense?
Is it somewhere already done. If yes, please help get me to the right place.

**Comment by honnibal** [comment]
The easiest way to run this sort of experiment will be to build from source and hack at the source code. Then once you know whether your ideas work, you can implement them more cleanly using the API.

The function you're looking for is in `spacy/_ml.py`, the `Tok2Vec` function, specifically. This function is called by `spacy/syntax/nn_parser.py::Parser.Model`, to define the shape of the neural network. The `Tok2Vec` function defines the embedding and CNN layers of the network; the input to these layers is a list of docs, and the output is a list of 2d tensors.

You'll want to create embedding tables for the attributes you're interested in. You can see how this is done in the current implementation, as we already concatenate the embeddings of several features (the normalized form, the prefix, suffix and word shape.

Specifically, we have:

```python
    cols = [ID, NORM, PREFIX, SUFFIX, SHAPE, ORTH]
    with Model.define_operators({'>>': chain, '|': concatenate, '**': clone,
                                 '+': add, '*': reapply}):
        norm = HashEmbed(width, embed_size, column=cols.index(NORM),
                         name='embed_norm')
        prefix = HashEmbed(width, embed_size//2, column=cols.index(PREFIX),
                           name='embed_prefix')
        suffix = HashEmbed(width, embed_size//2, column=cols.index(SUFFIX),
                           name='embed_suffix')
        shape = HashEmbed(width, embed_size//2, column=cols.index(SHAPE),
                          name='embed_shape')
        if pretrained_vectors is not None:
            glove = StaticVectors(pretrained_vectors, width, column=cols.index(ID))

            embed = uniqued(
                (glove | norm | prefix | suffix | shape)
                >> LN(Maxout(width, width*5, pieces=3)), column=cols.index(ORTH))
        else:
            embed = uniqued(
                (norm | prefix | suffix | shape)
                >> LN(Maxout(width, width*4, pieces=3)), column=cols.index(ORTH))
```

You need to make five changes:

1. Append a new attribute to `cols`, e.g. `TAG` or `DEP`
2. Create a new embedding table, e.g. `tag = HashEmbed(width, embed_size, column=cols.index(TAG), name='embed_tag')
3. Remove the `uniqued()` transform from the embedding function. This transform finds the unique rows in the feature table, indexed by the ORTH key, and then only embeds once per word type. This works because the features I'm using are all fully determined by the lexical type. Yours won't be --- so you need to remove this.
4. Add your embedding table to the concatenated feature vector
5. Change the input width of the hidden layer that's mixing the concatenated features.

Let me know if you get interesting results! I'd guess this won't help much.

**Comment by rbhambriiit** [comment]
Thanks for the detailed info. I am hoping to try it in near future.

I already have an NER model doing fairly good using the default spacy models - I am hoping it make it bit better by augmenting features.

One major motivation behind this is that some of my entity types are noun chunks and come up straightly available as noun chunks when I push my sentences to displaCy dependency parser. Problem is how to segregate the noun chunks which are named-entities of my interest vs. noise.  I just want to inject some of those dep. parser/POS tag features, hoping to teach the NER model some of this stuff i,e, leveraging the ner+parser capabilities. 

I also checked out crfsuite on sklearn for NER task and found that they also use POS tags. I also wonder why SpaCy does not need these basic features for NER. 
(I read that the parser layer does somehow rely on the tagger layer, but for NER I could not find that)


Secondly, I am also inspired by: 
https://github.com/explosion/spaCy/blob/master/examples/information_extraction/entity_relations.py
where we observe that Named entities and dependency parser come interact to get meaningful info.

**Comment by honnibal** [comment]
Fair enough!

spaCy 1 used POS tags in the NER model as well. When you have a linear model POS tags are very nice, because you're very keen to reduce the sparsity of the feature set. You can't make everything a trigram of words; you have to summarise. POS tags and the Brown cluster IDs are a good way of doing that.

With a neural network the situation is a bit different. The POS tagger sees a lexical representation that's pretty much exactly the same as the NER's lexical representation. Given that, there's not so much motivation to use the POS tags: the NER should be able to learn just the same thing, except without the clipping to a single ID (which loses information).

**Comment by bratao** [comment]
My task is to extract/segment information from a semi structured text. But not only the text is important some "external features" are also important.

For example, imagine that I want to segment this text about GitHub projects in Category, Project name, URl and description.

![image](https://user-images.githubusercontent.com/1090152/58190397-d14c0380-7c92-11e9-91e2-c1b9dd62737a.png)

I utilize an BIO scheme to tag each html token as a category. 

Where
token=NLP start_of_p=True bold=True center=True B-Category
token=Projects start_of_p=False bold=True center=True I-Category
token=Project start_of_p=True bold=True center=False B-Project-name
token=Name start_of_p=False bold=True center=False I-Project-name
token=: start_of_p=False bold=False center=False I-Project-name

The final result is something like:

![image](https://user-images.githubusercontent.com/1090152/58190447-eaed4b00-7c92-11e9-9279-f331324c828b.png)

Pay attention that some features are important such as: Text formatting (italic, bold, centered), position in text and more...

This would be a scenario where Spacy can work with?
I can see a lot of moviment around Tok2Vec. But looking at https://github.com/explosion/spaCy/blob/87fcf3141c16535b3a58edc11f799589e97fa6f1/spacy/_ml.py#L322 , apparently it still glove | norm | prefix | suffix | shape only.

There is anyway of using those custom external features for training Spacy?

Thank you!

**Comment by LarsNeR** [comment]
I have seen #4518 was merged some months ago. Are there any updates to this issue?

**Comment by adrianeboyd** [comment]
There will be a lot of improvements in spacy v3, which will use the new config system from thinc v8 (https://thinc.ai/docs/usage-config). You can see a preview in the `develop` branch of how you will be able to register components and write train configs: https://github.com/explosion/spaCy/tree/develop/spacy/ml/models

**Comment by adrianeboyd** [comment]
Please try out the new `spacy-nightly` release to see how the new config system makes it easier to specify the features for a particular model: https://nightly.spacy.io

Here's an example of what this looks like in the config:

```ini
[components.ner.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = 128
attrs = ["LOWER","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = false
```

You can define the features (if the parser is run first, you can add `DEP`, for instance). The `rows` is the number of rows in the embedding table (`embed_size` from the v2 code).

You can also more easily write your own custom embedding functions, see: https://nightly.spacy.io/usage/training

Please feel free to open a new issue if you run into problems!

**Comment by bratao** [comment]
@adrianeboyd I´m so excited to try it. 
It is possible to , for example, add a custom attribute such as is_bold for each token, and train a NER to use this information in training/prediction?
It is possible to do this by code, or just using the config pipeline?

Thank you so much!!!
